### PR TITLE
Add struct initialization to Spicy

### LINKED ISSUE #842: Add struct type to Spicy
Spicy doesn't have a `struct` type currently, which would be nice to have for maintaining custom data. 

There are two work-arounds available: (1) use a tuple with named elements, and (2) use a `unit` with only `var` elements. Option (1) gets cumbersome, and potentially inefficient, with increasing number of fields; and option (2) triggers all the parsing machinery and isn't nice syntax-wise either.

HILTI already has a `struct` type, I think all we need to do is expose that in Spicy.

On Zeek-side, structs should translate into Zeek records similar to what we do for tuples.

### PR DESCRIPTION
**Pull request by bbannier**
Closes #842.

### CODE REVIEW

File: spicy/toolchain/src/compiler/parser/parser.yy [hunk 1]
```diff
@@ -1027,7 +1027,7 @@ struct_       : '[' struct_elems ']'             { $$ = hilti::ctor::Struct(std:
 struct_elems  : struct_elems ',' struct_elem     { $$ = std::move($1); $$.push_back($3); }
               | struct_elem                      { $$ = std::vector<hilti::ctor::struct_::Field>{ std::move($1) }; }
 
-struct_elem   : '$' local_id  '=' expr           { $$ = hilti::ctor::struct_::Field(std::move($2), std::move($4)); }
+struct_elem   : DOLLAR_IDENT '=' expr            { $$ = hilti::ctor::struct_::Field(hilti::ID(std::move($1)), std::move($3)); }
```
[the remark this answers is not in the document]
Comment (bbannier) [reply spicy/toolchain/src/compiler/parser/parser.yy:1030]: Is it acceptable to merge this as is, or do we need to add the validation rejecting this first @rsmmr?

**Review by rsmmr — COMMENTED** [review]
Ok, let's merge it as it. However, we're making structs a supported Spicy type now, so we'll need to document it, too, at https://docs.zeek.org/projects/spicy/en/latest/programming/language/types.html

File: spicy/toolchain/src/compiler/parser/parser.yy (hunk 1 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply spicy/toolchain/src/compiler/parser/parser.yy:1030]: Still need to document this, https://github.com/zeek/spicy/pull/1039#pullrequestreview-798576992.

[the remark this answers is not in the document]
Comment (bbannier) [reply spicy/toolchain/src/compiler/parser/parser.yy:1030]: Added documentation, filed #1043 to reject the initialization w/o clear type.

File: doc/programming/language/types.rst [hunk 2]
```diff
@@ -439,6 +439,25 @@ set. Internally, Spicy stores them as UTF-8.
 
 .. include:: /autogen/types/string.rst
 
+.. _type_struct:
+
+Struct
+------
+
```
Comment (rsmmr) [inline doc/programming/language/types.rst:446]: Please add at least a couple of sentences of an introductory summary here (see the other types for examples).

File: doc/programming/language/types.rst [hunk 3]
```diff
@@ -497,9 +516,8 @@ Unit
 .. rubric:: Constants
 
 - Spicy doesn't support unit constants, but you can initialize unit
-  instances through coercion from a list expression: ``my_unit =
-  [$FIELD_1 = X_1, $FIELD_N = X_N, ...]`` where ``FIELD_I`` is the
-  label of a corresponding field in ``my_unit``'s type.
+  instances through coercion from a ``struct`` initializer, see
+  :ref:`type_struct`.
 
   .. todo:: This isn't available in Spicy yet (:issue:`1036`).
```
Comment (rsmmr) [inline doc/programming/language/types.rst:522]:
```suggestion
  .. todo:: This initialization isn't actually available yet (:issue:`1036`).
```

**Review by rsmmr — APPROVED** [review]

File: doc/programming/language/types.rst (hunk 2 above)
Comment (bbannier) [reply doc/programming/language/types.rst:446]: I'll add

> A structs is a heterogeneous container of a fixed, ordered set of named values similar to a Tuple.

**Review by rsmmr — APPROVED** [review]

**Pull request merged**